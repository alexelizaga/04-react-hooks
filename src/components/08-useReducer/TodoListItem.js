import React from 'react';
import PropTypes from 'prop-types';

export const TodoListItem = ({ index, todo, handleToggle, handleDelete }) => {
  return (
    <li key={todo.id} className="list-group-item">
      <p
        className={`${todo.done && "complete"}`}
        onClick={() => handleToggle(todo.id)}
      >
        {index + 1}. {todo.desc}
      </p>
      <button
        className="btn btn-danger"
        onClick={() => handleDelete(todo.id)}
      >
        Borrar
      </button>
    </li>
  )
};

TodoListItem.propTypes = {
  index: PropTypes.number.isRequired,
  todo: PropTypes.object.isRequired,
  handleToggle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
}
