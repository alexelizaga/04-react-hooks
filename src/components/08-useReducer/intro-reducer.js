
const initialState = [{
  id: 1,
  todo: 'Comprar pan',
  done: false,
}];

const todoReducer = ( state = initialState, action ) => {
  switch (action?.type) {
    case 'add':
      return [ ...state, action.payload ]
  
    default:
      return state;
  }
}
let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: 'Comprar leche',
  done: false,
}

const action = { type: 'add', payload: newTodo }

todos = todoReducer( todos, action );






console.log(todos);