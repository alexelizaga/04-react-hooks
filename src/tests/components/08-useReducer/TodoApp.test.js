import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';

import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Test <TodoApp />', () => {

  Storage.prototype.setItem = jest.fn( ()=>{} );

  test('should render correly', () => {
    const tree = renderer.create( <TodoApp />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should add a ToDo', () => {
    render(<TodoApp />);

    const h1 = screen.getByTestId('TodoAppTitle');
    expect(h1.textContent).toBe('TodoApp (0)');

    expect(localStorage.setItem).toBeCalledTimes(1);
  });

})