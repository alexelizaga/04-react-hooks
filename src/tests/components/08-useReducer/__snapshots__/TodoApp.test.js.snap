// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <TodoApp /> should render correly 1`] = `
<div>
  <h1
    data-testid="TodoAppTitle"
  >
    TodoApp (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <ul
        className="list-group list-group-flush"
      />
    </div>
    <div
      className="col-5"
    >
      <h4>
        Agregar TODO
      </h4>
      <hr />
      <form
        data-testid="TodoAddForm"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          data-testid="TodoAddInput"
          name="description"
          onChange={[Function]}
          placeholder="Aprender ..."
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1 btn-block"
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>
</div>
`;
